File: 2381-Shifting Letters II.py
class Solution(object):
    def shiftingLetters(self, s, shifts):
        """
        :type s: str
        :type shifts: List[List[int]]
        :rtype: str
        """
        dp = [0]*len(s)
        
        for start, end, direction in shifts:
            if direction == 0:
                dp[start] -= 1
                if end + 1 <= len(s)-1:
                    dp[end+1] += 1
            else:
                dp[start] += 1
                if end+1 <= len(s)-1:
                    dp[end+1] -= 1
        
        runningVal = 0
        
        res = []
        for i in range(len(dp)):
            runningVal += dp[i]
            curr = ord(s[i]) - ord('a')
            diff = (runningVal%26 + 26 + curr)%26
            res.append(chr(diff+ord('a')))
        
        return "".join(res)

File: C++/2381-Shifting Letters II.cpp
class Solution {
public:
    string shiftingLetters(string s, vector<vector<int>>& shifts) {
        int N = s.size();
        vector<int> dp(N+1, 0);
        for (auto &s:shifts) {
            int u = s[0], v = s[1], w = s[2];
            if (w == 0) {
                dp[u]--;
                dp[v+1]++;
            }  else {
                dp[u]++;
                dp[v+1]--;
            }
        }

        int running = 0;
        for (int i=0; i<N; i++) {
            running += dp[i];
            s[i] = ( (s[i] - 'a' + running)%26 + 26)%26 + 'a';
        }

        return s;
    }
};